test/core/event_engine/fuzzing_event_engine/fuzzing_event_engine.h

```diff
@@ -224,6 +233,48 @@ class FuzzingEventEngine : public EventEngine {
     const int index_;
   };
 
+  // Implementation of a MockEndpoint.
+  // Writes made using this Endpoint always succeed and are discarded. Arbitrary
+  // data can be injected into the Mock Endpoint to be read.
+  // MockEndpointActions is a list of actions which specifies the set of bytes
+  // that need to be read and the time at which these bytes need to be read.
+  // The first element of each action indicates the relative delay from the
+  // previous action. The second element of each action specifies the set of
+  // bytes to be read when the absolute time advances and triggers the action.
+  class MockEndpoint final : public Endpoint {
+   public:
+    MockEndpoint(int local_port, int peer_port,
+                 const MockEndpointActions& actions) {
+      local_address_ = *URIToResolvedAddress(
+          absl::StrFormat("ipv4:127.0.0.1:%d", local_port));
+      peer_address_ = *URIToResolvedAddress(
+          absl::StrFormat("ipv4:127.0.0.1:%d", peer_port));
+      Init(actions);
+    }
+    ~MockEndpoint() override = default;
+
+    void Init(const MockEndpointActions& actions);
+
+    // A call to MockEndpoint::Read will query the current time and determine
+    // the set of bytes to be read based on passed MockEndpointActions.
+    bool Read(absl::AnyInvocable<void(absl::Status)> on_read,
+              SliceBuffer* buffer, const ReadArgs* args) override;
+    bool Write(absl::AnyInvocable<void(absl::Status)> on_writable,
+               SliceBuffer* data, const WriteArgs* args) override;
+
+    const ResolvedAddress& GetPeerAddress() const override {
+      return peer_address_;
+    }
+    const ResolvedAddress& GetLocalAddress() const override {
+      return local_address_;
+    }
+
+   private:
+    ResolvedAddress local_address_;
+    ResolvedAddress peer_address_;
+    std::vector<std::tuple<Time, std::string>> actions_by_time_;
```

Comment:
I think there's a simpler implementation available:

If we have a `std::vector<uint8_t> bytes_available_to_read_;` here then our initialization loop can look like:

```
for (const MockEndpointAction& action : actions) {
  event_engine->RunAfter(action.time, [this]() {
    bytes_available_to_read_.extend(action.bytes);
  });
}
```

and `Read` itself becomes a fairly trivial copy.

(ok sans delayed reads, but I think they can fall out fairly easily with an `on_complete_` member also, since `AnyInvocable` is nullable)